raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -337,26 +342,39 @@ private VoteResponseData handleVoteRequest(VoteRequestData request) throws IOExc
         }
     }
 
-    private void handleVoteResponse(int remoteNodeId, VoteResponseData response) throws IOException {
-        if (quorum.isLeader()) {
-            logger.debug("Ignoring vote response {} since we already became leader for epoch {}",
+    private Optional<Errors> handleVoteResponse(
+        RaftResponse.Inbound responseMetadata
+    ) throws  IOException {
+        int remoteNodeId = responseMetadata.sourceId();
+        VoteResponseData response = (VoteResponseData) responseMetadata.data;
+        Errors error = Errors.forCode(response.errorCode());
+
+        if (handleNonMatchingResponseLeaderAndEpoch(response.leaderEpoch(), optionalLeaderId(response.leaderId()))) {
+            return Optional.of(error);
+        } else if (error == Errors.NONE) {
+            if (quorum.isLeader()) {
+                logger.debug("Ignoring vote response {} since we already became leader for epoch {}",
                     response, quorum.epoch());
-        } else if (quorum.isFollower()) {
-            logger.debug("Ignoring vote response {} since we are now a follower for epoch {}",
+            } else if (quorum.isFollower()) {
+                logger.debug("Ignoring vote response {} since we are now a follower for epoch {}",
                     response, quorum.epoch());
-        } else {
-            CandidateState state = quorum.candidateStateOrThrow();
-            if (response.voteGranted()) {
-                state.recordGrantedVote(remoteNodeId);
-                maybeBecomeLeader(state);
             } else {
-                state.recordRejectedVote(remoteNodeId);
-                if (state.isVoteRejected()) {
-                    logger.info("Insufficient remaining votes to become leader (rejected by {}). " +
-                            "We will await expiration of election timeout before retrying",
-                        state.rejectingVoters());
+                CandidateState state = quorum.candidateStateOrThrow();
+                if (response.voteGranted()) {
+                    state.recordGrantedVote(remoteNodeId);
+                    maybeBecomeLeader(state);
+                } else {
+                    state.recordRejectedVote(remoteNodeId);
+                    if (state.isVoteRejected()) {
```

Comment:
Thinking about this a bit more after #309 , it's a bit different between "adding the jitter to the election timeout" v.s. "backoff jitter time before issuing a new election" since in the first case (our current impl) we could start sending the vote request right after, where as in the latter case, we first wait for jitter, then start sending the vote request.

So back to this example, I think we should reset the electionTimer to `jitter` so that we can become a candidate after jitter time instead of letting the whole election time to be elapsed. And in `becomeCandidate` we should not use `election + jitter` but should always use `election` (I think in the original paper we do not need a random jitter for election itself, just the jitter for backing off?)

Reply:
Let's discuss this separately. I think my assumption has been that jitter is small relative to the election timeout. If that is not the conventional approach, then we should change it.